narwhal/consensus/src/consensus.rs

```diff
@@ -31,6 +35,114 @@ pub mod consensus_tests;
 /// The representation of the DAG in memory.
 pub type Dag = BTreeMap<Round, HashMap<AuthorityIdentifier, (CertificateDigest, Certificate)>>;
 
+#[derive(Default)]
+pub struct LeaderSwapTable {
+    pub good_nodes: Vec<AuthorityIdentifier>,
+    pub bad_nodes: HashSet<AuthorityIdentifier>,
+}
+
+impl LeaderSwapTable {
+    // constructs a new table based on the provided reputation scores.
+    pub fn new(committee: &Committee, reputation_scores: ReputationScores) -> Self {
+        assert!(reputation_scores.final_of_schedule, "Only reputation scores that have been calculated on the end of a schedule are accepted");
+
+        let good_nodes = Self::good_nodes(&reputation_scores, committee);
+        let bad_nodes = Self::bad_nodes(&reputation_scores, committee);
+
+        debug!("Reputation scores: {:?}", reputation_scores);
+        debug!("Good nodes: {:?}", good_nodes);
+        debug!("Bad nodes: {:?}", bad_nodes);
+
+        Self {
+            good_nodes,
+            bad_nodes,
+        }
+    }
+
+    /// Checks whether the provided leader is a bad performer and needs to be swapped for the schedule
+    /// with a good performer. If not, then the method returns None. Otherwise the leader to swap with
+    /// is returned instead. The `leader_round` represents the DAG round on which the provided AuthorityIdentifier
+    /// is a leader on and is used as a seed to random function in order to calculate the good node that
+    /// will swap for that round the bad node. We are intentionally not doing weighted randomness as
+    /// we want to give to all the good nodes equal opportunity to get swapped with bad nodes and not
+    /// have one node with enough stake dominating on the final schedule.
+    pub fn swap(
+        &self,
+        leader: &AuthorityIdentifier,
+        leader_round: Round,
+    ) -> Option<AuthorityIdentifier> {
+        if self.bad_nodes.contains(leader) {
+            let mut seed_bytes = [0u8; 32];
+            seed_bytes[32 - 8..].copy_from_slice(&leader_round.to_le_bytes());
+            let mut rng = StdRng::from_seed(seed_bytes);
+
+            let good_node = self
+                .good_nodes
+                .choose(&mut rng)
+                .expect("There should be at least one good node available");
+
+            trace!(
+                "Swapping bad leader {} -> {} for round {}",
+                leader,
+                good_node,
+                leader_round
+            );
+
+            return Some(*good_node);
+        }
+        None
+    }
+
+    /// Retrieves the f by stake nodes with the best scores.
```

Comment:
Hhm to be honest I've split those for that exact reason, readability and keeping the `new` method more concise. If you believe that it's that bad I can change it, otherwise I would prefer to keep them as is for now.

Reply:
I will share my reasoning but you can choose what you think is best!

My reasoning is that when I go to read this code or perhaps debug this code at any point I have to jump around to another part of the file to figure out how `good_nodes` or `bad_nodes` is calculated. If this was abstracting away a large amount of complicated code then I would totally agree with moving it into another method, but because it is essentially a wrapper for `retrieve_first_f_nodes()` it just becomes a level of indirection that IMO makes the code a little bit harder to navigate.

one other very small win is you only have to call `reputation_scores.authorities_by_score_desc();` once if you inline the changes 😄